### ISSUE TITLE
IO.copy_stream keep pos of destination to beginning of file

### ISSUE THREAD

**Issue by HoneyryderChuck** [oracle/truffleruby#3280]
This is the minimum repro I could make:

```ruby
require "stringio"
require "tempfile"
a = StringIO.new
a << ("a"*100000)
b = Tempfile.new("test", encoding: Encoding::BINARY, mode: File::RDWR)
a.rewind
IO.copy_stream(a, b)
puts b.pos
```

Using ruby 3.2, the output is 100000, whereas in truffleruby, output is 0.

version was `truffleruby 22.3.1, like ruby 3.0.3, GraalVM CE Native [x86_64-linux]` (I use truffleruby for tests in docker, there is no v23 image available yet).

**Comment by eregon** [comment]
Thank you for the report.
This bug is due to https://github.com/oracle/truffleruby/blob/c6e61627c539910ed6c3d165763bf255f3c9bc41/src/main/ruby/truffleruby/core/io.rb#L368-L370

A Tempfile is not an IO or a File, it's a separate class using delegation:
```
irb(main):006:0> Tempfile.ancestors
=> [Tempfile, #<Class:0x00007f9b2de25e80>, Delegator, #<Module:0x00007f9b2de0a338>, BasicObject]
```

And so the coercion tries `to_path` which does exist on on Tempfile and Delegates to File#to_path, and therefore it creates another IO object.
I will check what's the actual logic in CRuby, maybe we shouldn't try `to_path` there or maybe there is a special case for Tempfile.

**Comment by eregon** [comment]
https://github.com/ruby/ruby/blob/2d6067dc7d02c5942da6636bba8b1629065aecf0/io.c#L13160 is the conversion logic in CRuby.
So it check if dst is ARGF or is not (an IO or a String or a #to_path), then dst_fptr = NULL.
And if that doesn't hold (Tempfile has to_path) then it tries rb_io_check_io which is #to_io first, then if not an IO tries FilePathValue (#to_path) and otherwise uses it as-is as an IO.

**Comment by eregon** [comment]
Fix in https://github.com/oracle/truffleruby/pull/3282

**Comment by HoneyryderChuck** [comment]
Thx for the prompt fix @eregon 🙏

Unrelated,  but when can one expect docker truffleruby images targeting v23?

**Comment by eregon** [comment]
@HoneyryderChuck The Docker images for 23.0 and 23.1 are already available: https://github.com/graalvm/container/blob/master/truffleruby-community/README.md

**Comment by HoneyryderChuck** [comment]
Oh, i was still using `ghcr.io/graalvm/truffleruby:latest`, didn't know about the change. will upate CI builds accoringly, thx!

**Comment by eregon** [comment]
This fix will be in the 23.1.2 release which will be available on January 23, 2024.
